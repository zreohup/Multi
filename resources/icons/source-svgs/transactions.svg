<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M22.6788 13.4894C22.2506 13.0434 21.5505 13.0434 21.1223 13.4894L18.6014 16.1086V5.14372C18.6014 4.51467 18.106 4 17.5005 4C16.894 4 16.3997 4.51467 16.3997 5.14372V16.1086L13.8777 13.4894C13.4495 13.0445 12.7494 13.0445 12.3212 13.4894C11.8929 13.9344 11.8929 14.6618 12.3212 15.1067L16.7223 19.6781C16.7861 19.7467 16.8676 19.7833 16.9424 19.8302C16.9887 19.8588 17.0261 19.9 17.0767 19.9229C17.1219 19.9435 17.1725 19.9435 17.2198 19.9549C17.4532 20.0201 17.6998 20.0189 17.9244 19.9229C17.9728 19.9011 18.0091 19.8611 18.0543 19.8325C18.1313 19.7856 18.2117 19.7479 18.2788 19.6781L22.6788 15.1067C23.1071 14.6618 23.1071 13.9344 22.6788 13.4894Z" fill="black"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M11.6788 10.5106C11.2506 10.9566 10.5505 10.9566 10.1223 10.5106L7.60138 7.89144V18.8563C7.60138 19.4853 7.106 20 6.50055 20C5.894 20 5.39972 19.4853 5.39972 18.8563V7.89144L2.87773 10.5106C2.44951 10.9555 1.74939 10.9555 1.32117 10.5106C0.892945 10.0656 0.892945 9.33824 1.32117 8.89333L5.72227 4.32189C5.78611 4.25326 5.86758 4.21666 5.94243 4.16977C5.98867 4.14118 6.02609 4.1 6.07673 4.07713C6.12187 4.05654 6.1725 4.05654 6.21984 4.04511C6.45321 3.97991 6.6998 3.98106 6.92437 4.07713C6.9728 4.09886 7.00913 4.13889 7.05427 4.16748C7.13132 4.21438 7.21168 4.25212 7.27883 4.32189L11.6788 8.89333C12.1071 9.33824 12.1071 10.0656 11.6788 10.5106Z" fill="black"/>
</svg>
