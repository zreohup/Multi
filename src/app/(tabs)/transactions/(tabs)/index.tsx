import { TxHistoryContainer } from '@/src/features/TxHistory'
import { View } from 'react-native'

export default function TransactionsScreen() {
  return (
    <View style={{ flex: 1 }}>
      <TxHistoryContainer />
    </View>
  )
}
