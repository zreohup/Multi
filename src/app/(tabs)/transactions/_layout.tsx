import { Stack } from 'expo-router'
import React from 'react'
import type { Route } from '@react-navigation/routers'
import { H2, View } from 'tamagui'
import { getFocusedRouteNameFromRoute } from '@react-navigation/native'

const getHeaderTitle = (route: Partial<Route<string>>) => {
  const routeName = getFocusedRouteNameFromRoute(route) ?? 'index'
  const name = {
    ['index']: 'Transactions',
    ['messages']: 'Messages',
  }[routeName]
  return name || 'Transactions'
}

export default function TransactionsLayout() {
  return (
    <Stack
      screenOptions={{
        headerLargeTitle: false,
        headerShadowVisible: false,
      }}
    >
      <Stack.Screen
        name="(tabs)"
        options={({ route }) => ({
          headerTitle: (props) => (
            <View
              style={{
                marginTop: 2,
                flex: 1,
                width: '100%',
              }}
            >
              <H2 fontWeight={600} {...props}>
                {getHeaderTitle(route)}
              </H2>
            </View>
          ),
        })}
      />
    </Stack>
  )
}
