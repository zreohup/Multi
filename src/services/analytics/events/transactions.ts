import { EventType, type AnalyticsLabel } from '../types'

const TRANSACTIONS_CATEGORY = 'transactions'

export const TRANSACTIONS_EVENTS = {
  CONFIRM: {
    eventName: EventType.TX_CONFIRMED,
    eventCategory: TRANSACTIONS_CATEGORY,
    eventAction: 'Confirm transaction',
    // eventLabel can be set when tracking the event using AnalyticsLabel
  },
  CREATE: {
    eventName: EventType.TX_CREATED,
    eventCategory: TRANSACTIONS_CATEGORY,
    eventAction: 'Create transaction',
    // eventLabel can be set when tracking the event using AnalyticsLabel
  },
  EXECUTE: {
    eventName: EventType.TX_EXECUTED,
    eventCategory: TRANSACTIONS_CATEGORY,
    eventAction: 'Execute transaction',
    // eventLabel can be set when tracking the event using AnalyticsLabel
  },
}

// Helper function to create a transaction confirm event with a specific label
export const createTxConfirmEvent = (label: AnalyticsLabel) => ({
  ...TRANSACTIONS_EVENTS.CONFIRM,
  eventLabel: label,
})

// Helper function to create a transaction create event with a specific label
export const createTxCreateEvent = (label: AnalyticsLabel) => ({
  ...TRANSACTIONS_EVENTS.CREATE,
  eventLabel: label,
})

// Helper function to create a transaction execute event with a specific label
export const createTxExecuteEvent = (label: AnalyticsLabel) => ({
  ...TRANSACTIONS_EVENTS.EXECUTE,
  eventLabel: label,
})
