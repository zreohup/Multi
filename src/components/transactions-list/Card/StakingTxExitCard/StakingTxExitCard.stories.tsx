import type { Meta, StoryObj } from '@storybook/react'
import { StakingTxExitCard } from '@/src/components/transactions-list/Card/StakingTxExitCard'
import { NativeStakingValidatorsExitTransactionInfo } from '@safe-global/store/gateway/AUTO_GENERATED/transactions'
import { TransactionInfoType } from '@safe-global/store/gateway/types'

// Mock data for NativeStakingValidatorsExitTransactionInfo
const mockStakingTxExitInfo: NativeStakingValidatorsExitTransactionInfo = {
  type: 'NativeStakingValidatorsExit' as TransactionInfoType.NATIVE_STAKING_VALIDATORS_EXIT,
  humanDescription: 'Exit staking validators',
  numValidators: 2,
  status: 'EXITING',
  estimatedExitTime: 1703980800, // Unix timestamp for 2023-12-31
  estimatedWithdrawalTime: 1704585600, // Unix timestamp for 2024-01-07
  value: '32000000000000000000', // 32 ETH in wei
  validators: ['0xvalidator1', '0xvalidator2'],
  tokenInfo: {
    address: '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE', // Common ETH placeholder address
    decimals: 18,
    logoUri: 'https://safe-transaction-assets.safe.global/chains/1/chain_logo.png',
    name: 'Ethereum',
    symbol: 'ETH',
    trusted: true,
  },
}

const meta: Meta<typeof StakingTxExitCard> = {
  title: 'TransactionsList/StakingTxExitCard',
  component: StakingTxExitCard,
  argTypes: {},
  args: {
    info: mockStakingTxExitInfo,
  },
}

export default meta
type Story = StoryObj<typeof StakingTxExitCard>

export const Default: Story = {}

export const SingleValidator: Story = {
  args: {
    info: {
      ...mockStakingTxExitInfo,
      numValidators: 1,
      validators: ['0xvalidator1'],
    },
  },
}

export const MultipleValidators: Story = {
  args: {
    info: {
      ...mockStakingTxExitInfo,
      numValidators: 5,
      validators: ['0xvalidator1', '0xvalidator2', '0xvalidator3', '0xvalidator4', '0xvalidator5'],
      value: '160000000000000000000', // 160 ETH in wei (5 validators * 32 ETH)
    },
  },
}

export const DifferentToken: Story = {
  args: {
    info: {
      ...mockStakingTxExitInfo,
      tokenInfo: {
        address: '0x1111111111',
        decimals: 18,
        logoUri:
          'https://safe-transaction-assets.safe.global/tokens/logos/0x5aFE3855358E112B5647B952709E6165e1c1eEEe.png',
        name: 'SafeToken',
        symbol: 'SAFE',
        trusted: true,
      },
    },
  },
}
