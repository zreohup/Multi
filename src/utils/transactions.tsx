import { GroupedTxs } from '@/src/features/TxHistory/utils'
import {
  isTransferTxInfo,
  isAddSignerTxInfo,
  isMultiSendTxInfo,
  isCustomTxInfo,
  isERC721Transfer,
  isRemoveSignerTxInfo,
  isOrderTxInfo,
  isVaultDepositTxInfo,
  isVaultRedeemTxInfo,
  isStakingTxDepositInfo,
  isStakingTxExitInfo,
  isStakingTxWithdrawInfo,
  isCancellationTxInfo,
  isBridgeOrderTxInfo,
  isLifiSwapTxInfo,
} from '@/src/utils/transaction-guards'
import { Transaction } from '@safe-global/store/gateway/AUTO_GENERATED/transactions'
import { ETxType } from '../types/txType'

export const getTransactionType = ({ txInfo }: { txInfo: Transaction['txInfo'] }) => {
  if (isTransferTxInfo(txInfo)) {
    return ETxType.TOKEN_TRANSFER
  }

  if (isTransferTxInfo(txInfo) && isERC721Transfer(txInfo)) {
    return ETxType.NFT_TRANSFER
  }

  if (isAddSignerTxInfo(txInfo)) {
    return ETxType.ADD_SIGNER
  }

  if (isRemoveSignerTxInfo(txInfo)) {
    return ETxType.REMOVE_SIGNER
  }

  if (isCancellationTxInfo(txInfo)) {
    return ETxType.CANCEL_TX
  }

  if (isMultiSendTxInfo(txInfo) || isCustomTxInfo(txInfo)) {
    return ETxType.CONTRACT_INTERACTION
  }

  if (isOrderTxInfo(txInfo)) {
    return ETxType.SWAP_ORDER
  }

  if (isBridgeOrderTxInfo(txInfo)) {
    return ETxType.BRIDGE_ORDER
  }

  if (isLifiSwapTxInfo(txInfo)) {
    return ETxType.LIFI_SWAP
  }

  if (isStakingTxDepositInfo(txInfo)) {
    return ETxType.STAKE_DEPOSIT
  }

  if (isStakingTxExitInfo(txInfo)) {
    return ETxType.STAKE_WITHDRAW_REQUEST
  }

  if (isStakingTxWithdrawInfo(txInfo)) {
    return ETxType.STAKE_EXIT
  }

  if (isVaultDepositTxInfo(txInfo)) {
    return ETxType.VAULT_DEPOSIT
  }

  if (isVaultRedeemTxInfo(txInfo)) {
    return ETxType.VAULT_REDEEM
  }

  return null
}

export const groupBulkTxs = <T extends { type: string; transaction?: Transaction }>(
  list: GroupedTxs<T>,
): GroupedTxs<T> => {
  const grouped = list
    .reduce<GroupedTxs<T>>((resultItems, item) => {
      if (Array.isArray(item) || item.type !== 'TRANSACTION') {
        return resultItems.concat([item])
      }
      const currentTxHash = item.transaction?.txHash

      const prevItem = resultItems[resultItems.length - 1]
      if (!Array.isArray(prevItem)) {
        return resultItems.concat([[item]])
      }
      const prevTxHash = prevItem[0]?.transaction?.txHash

      if (currentTxHash && currentTxHash === prevTxHash) {
        prevItem.push(item)
        return resultItems
      }

      return resultItems.concat([[item]])
    }, [])
    .map((item) => (Array.isArray(item) && item.length === 1 ? item[0] : item))

  return grouped
}
