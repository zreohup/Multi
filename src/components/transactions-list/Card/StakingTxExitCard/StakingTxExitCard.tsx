import type { NativeStakingValidatorsExitTransactionInfo } from '@safe-global/store/gateway/AUTO_GENERATED/transactions'
import { SafeListItem } from '@/src/components/SafeListItem'
import { TokenIcon } from '@/src/components/TokenIcon'
import { maybePlural } from '@safe-global/utils/utils/formatters'
import { Text } from 'tamagui'

interface StakingTxExitCardProps {
  info: NativeStakingValidatorsExitTransactionInfo
  onPress: () => void
}

export const StakingTxExitCard = ({ info, onPress }: StakingTxExitCardProps) => {
  return (
    <SafeListItem
      label={`Withdraw`}
      icon="transaction-stake"
      type={'Stake'}
      onPress={onPress}
      rightNode={
        <Text color="$color" fontWeight={600} textAlign="right">
          {info.numValidators} Validator{maybePlural(info.numValidators)}
        </Text>
      }
      leftNode={<TokenIcon logoUri={info.tokenInfo.logoUri} accessibilityLabel={info.tokenInfo.symbol} />}
    />
  )
}
