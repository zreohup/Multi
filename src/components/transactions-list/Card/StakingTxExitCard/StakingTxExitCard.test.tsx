import React from 'react'
import { render, fireEvent } from '@/src/tests/test-utils'
import { StakingTxExitCard } from './StakingTxExitCard'
import { NativeStakingValidatorsExitTransactionInfo } from '@safe-global/store/gateway/AUTO_GENERATED/transactions'

const mockInfo: NativeStakingValidatorsExitTransactionInfo = {
  type: 'NativeStakingValidatorsExit',
  humanDescription: null,
  status: 'ACTIVE',
  estimatedExitTime: 1234567890,
  estimatedWithdrawalTime: 1234567890,
  value: '32000000000000000000',
  numValidators: 1,
  tokenInfo: {
    address: '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE',
    decimals: 18,
    logoUri: 'https://safe-transaction-assets.safe.global/chains/1/chain_logo.png',
    name: 'Ethereum',
    symbol: 'ETH',
    trusted: true,
  },
  validators: ['0x123...abc'],
}

const mockOnPress = jest.fn()

describe('StakingTxExitCard', () => {
  beforeEach(() => {
    mockOnPress.mockClear()
  })

  it('matches snapshot', () => {
    const { toJSON } = render(<StakingTxExitCard info={mockInfo} onPress={mockOnPress} />)
    expect(toJSON()).toMatchSnapshot()
  })

  it('renders correctly', () => {
    const screen = render(<StakingTxExitCard info={mockInfo} onPress={mockOnPress} />)
    expect(screen.getByText('Withdraw')).toBeTruthy()
    expect(screen.getByText('1 Validator')).toBeTruthy()
  })

  it('renders multiple validators correctly', () => {
    const singleValidatorInfo = {
      ...mockInfo,
      numValidators: 3,
    }

    const screen = render(<StakingTxExitCard info={singleValidatorInfo} onPress={mockOnPress} />)
    expect(screen.getByText('Withdraw')).toBeTruthy()
    expect(screen.getByText('3 Validators')).toBeTruthy()
  })

  it('calls onPress when pressed', () => {
    const screen = render(<StakingTxExitCard info={mockInfo} onPress={mockOnPress} />)

    const card = screen.getByText('Withdraw')

    fireEvent.press(card)

    expect(mockOnPress).toHaveBeenCalledTimes(1)
  })
})
