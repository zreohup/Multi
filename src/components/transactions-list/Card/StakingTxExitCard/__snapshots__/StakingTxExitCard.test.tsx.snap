// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StakingTxExitCard matches snapshot 1`] = `
<View>
  <View
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "flex-start",
        "backgroundColor": "#FFFFFF",
        "borderBottomLeftRadius": 7,
        "borderBottomRightRadius": 7,
        "borderTopLeftRadius": 7,
        "borderTopRightRadius": 7,
        "flexDirection": "column",
        "flexWrap": "wrap",
        "gap": 12,
        "justifyContent": "flex-start",
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 12,
            "maxWidth": "55%",
          }
        }
      >
        <View
          style={
            {
              "width": 40,
            }
          }
        >
          <View
            style={
              {
                "position": "absolute",
                "right": -10,
                "top": -10,
                "zIndex": 1,
              }
            }
          />
          <View
            style={
              {
                "alignItems": "center",
                "borderBottomLeftRadius": 100000,
                "borderBottomRightRadius": 100000,
                "borderTopLeftRadius": 100000,
                "borderTopRightRadius": 100000,
                "flexDirection": "column",
                "height": 40,
                "justifyContent": "center",
                "maxHeight": 40,
                "maxWidth": 40,
                "minHeight": 40,
                "minWidth": 40,
                "overflow": "hidden",
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
                "position": "relative",
                "width": 40,
              }
            }
          >
            <View
              style={
                {
                  "bottom": 0,
                  "flexDirection": "column",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "zIndex": 1,
                }
              }
            >
              <Image
                accessibilityLabel="ETH"
                fullscreen={true}
                onError={[Function]}
                onLoad={[Function]}
                source={
                  {
                    "uri": "https://safe-transaction-assets.safe.global/chains/1/chain_logo.png",
                  }
                }
                style={
                  {
                    "backgroundColor": "#121312",
                    "height": 40,
                    "width": 40,
                  }
                }
                testID="logo-image"
              />
            </View>
            <View
              style={
                {
                  "backgroundColor": "#EEEFF0",
                  "bottom": 0,
                  "flexDirection": "column",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "zIndex": 0,
                }
              }
            >
              <View
                style={
                  {
                    "backgroundColor": "#EEEFF0",
                    "borderBottomLeftRadius": 100,
                    "borderBottomRightRadius": 100,
                    "borderTopLeftRadius": 100,
                    "borderTopRightRadius": 100,
                    "paddingBottom": 8,
                    "paddingLeft": 8,
                    "paddingRight": 8,
                    "paddingTop": 8,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    [
                      {
                        "color": "#A1A3A7",
                        "fontSize": 24,
                      },
                      undefined,
                      {
                        "fontFamily": "SafeIcons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      {},
                    ]
                  }
                  testID="logo-fallback-icon"
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View>
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "gap": 4,
                "marginBottom": 4,
              }
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#A1A3A7",
                    "fontSize": 10,
                  },
                  undefined,
                  {
                    "fontFamily": "SafeIcons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
              testID="safe-list-transaction-stake-icon"
            >
              
            </Text>
            <Text
              style={
                {
                  "color": "#A1A3A7",
                  "fontFamily": "DM Sans",
                  "fontSize": 13,
                  "marginBottom": 2,
                }
              }
              suppressHighlighting={true}
            >
              Stake
            </Text>
          </View>
          <Text
            style={
              {
                "color": "#121312",
                "fontFamily": "DMSans-SemiBold",
                "fontSize": 14,
              }
            }
            suppressHighlighting={true}
          >
            Withdraw
          </Text>
        </View>
      </View>
      <Text
        style={
          {
            "color": "#121312",
            "fontFamily": "DMSans-SemiBold",
            "textAlign": "right",
          }
        }
        suppressHighlighting={true}
      >
        1
         Validator
      </Text>
    </View>
  </View>
</View>
`;
